File: skyspark-scala/src/test/scala/com/github/dkanellis/skyspark/scala/api/algorithms/bnl/BnlAlgorithmTest.scala
package com.github.dkanellis.skyspark.scala.api.algorithms.bnl

import com.github.dkanellis.skyspark.scala.api.algorithms.Point
import com.github.dkanellis.skyspark.scala.test_utils.UnitSpec

class BnlAlgorithmTest extends UnitSpec {
  var bnlAlgorithm: BnlAlgorithm = _

  before { bnlAlgorithm = new BnlAlgorithm }

  "Without pre-comparison a set of points" should "compute the skylines" in {
    val points = Seq(
      Point(5.4, 4.4), Point(5.0, 4.1), Point(3.6, 9.0), Point(5.9, 4.0), Point(5.9, 4.6), Point(2.5, 7.3),
      Point(6.3, 3.5), Point(9.9, 4.1), Point(6.7, 3.3), Point(6.1, 3.4))
    val expectedSkylines = Seq(Point(5.0, 4.1), Point(5.9, 4.0), Point(2.5, 7.3), Point(6.7, 3.3), Point(6.1, 3.4))

    val actualSkylines = bnlAlgorithm.computeSkylinesWithoutPreComparison(points)

    actualSkylines should contain theSameElementsAs expectedSkylines
  }

  "Without pre-comparison a set of only skylines" should "have nothing changed" in {
    val points = Seq(Point(5.0, 4.1), Point(5.9, 4.0), Point(2.5, 7.3), Point(6.7, 3.3), Point(6.1, 3.4))

    val actualSkylines = bnlAlgorithm.computeSkylinesWithoutPreComparison(points)

    actualSkylines should contain theSameElementsAs points
  }

  "With pre-comparison a set of points with their flags" should "compute the skylines" in {
    val flagsPoints = Seq(
      (Flag(true, true), Point(5.9, 4.6)), (Flag(true, false), Point(5.0, 4.1)),
      (Flag(true, false), Point(5.9, 4.0)), (Flag(true, false), Point(6.7, 3.3)),
      (Flag(true, false), Point(6.1, 3.4)), (Flag(false, true), Point(2.5, 7.3)))
    val expectedSkylines = Seq(Point(5.0, 4.1), Point(5.9, 4.0), Point(2.5, 7.3), Point(6.7, 3.3), Point(6.1, 3.4))

    val actualSkylines = bnlAlgorithm.computeSkylinesWithPreComparison(flagsPoints)

    actualSkylines should contain theSameElementsAs expectedSkylines
  }

  "With pre-comparison a set of only skylines with their flags" should "have nothing changed" in {
    val flagsPoints = Seq(
      (Flag(true, false), Point(5.0, 4.1)), (Flag(true, false), Point(5.9, 4.0)), (Flag(true, false), Point(6.7, 3.3)),
      (Flag(true, false), Point(6.1, 3.4)), (Flag(false, true), Point(2.5, 7.3)))
    val expectedSkylines = Seq(Point(5.0, 4.1), Point(5.9, 4.0), Point(6.7, 3.3), Point(6.1, 3.4), Point(2.5, 7.3))

    val actualSkylines = bnlAlgorithm.computeSkylinesWithPreComparison(flagsPoints)

    actualSkylines should contain theSameElementsAs expectedSkylines
  }
}

File: skyspark-java/src/test/java/com/akikanellis/skyspark/api/algorithms/bnl/BnlAlgorithmTest.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import org.junit.Before;
import org.junit.Test;
import scala.Tuple2;

import java.util.Arrays;
import java.util.List;

import static com.akikanellis.skyspark.api.test_utils.assertions.MoreAssertions.assertThat;

public class BnlAlgorithmTest {
    private BnlAlgorithm bnl;

    @Before public void beforeEach() { bnl = new BnlAlgorithm(); }

    @Test public void computingSkylinesWithoutPreComparison_withPoints_returnsCorrectSkylines() {
        List<Point> points = Arrays.asList(
                new Point(5.4, 4.4), new Point(5.0, 4.1), new Point(3.6, 9.0),
                new Point(5.9, 4.0), new Point(5.9, 4.6), new Point(2.5, 7.3),
                new Point(6.3, 3.5), new Point(9.9, 4.1), new Point(6.7, 3.3), new Point(6.1, 3.4));
        List<Point> expectedSkylines = Arrays.asList(
                new Point(5.0, 4.1), new Point(5.9, 4.0),
                new Point(2.5, 7.3), new Point(6.7, 3.3), new Point(6.1, 3.4));

        Iterable<Point> actualSkylines = bnl.computeSkylinesWithoutPreComparison(points);

        assertThat(actualSkylines).containsOnlyElementsOf(expectedSkylines);
    }

    @Test public void computingSkylinesWithoutPreComparison_withOnlySkylines_returnsTheSamePoints() {
        List<Point> points = Arrays.asList(
                new Point(5.0, 4.1), new Point(5.9, 4.0), new Point(2.5, 7.3),
                new Point(6.7, 3.3), new Point(6.1, 3.4)
        );

        Iterable<Point> actualSkylines = bnl.computeSkylinesWithoutPreComparison(points);

        assertThat(actualSkylines).containsOnlyElementsOf(points);
    }

    @Test public void computingSkylinesWithPreComparison_withPoints_returnsCorrectSkylines() {
        List<Tuple2<Flag, Point>> flagsPoints = Arrays.asList(
                new Tuple2<>(new Flag(true, true), new Point(5.9, 4.6)),
                new Tuple2<>(new Flag(true, false), new Point(5.0, 4.1)),
                new Tuple2<>(new Flag(true, false), new Point(5.9, 4.0)),
                new Tuple2<>(new Flag(true, false), new Point(6.7, 3.3)),
                new Tuple2<>(new Flag(true, false), new Point(6.1, 3.4)),
                new Tuple2<>(new Flag(false, true), new Point(2.5, 7.3)));
        List<Point> expectedSkylines = Arrays.asList(
                new Point(5.0, 4.1), new Point(5.9, 4.0),
                new Point(2.5, 7.3), new Point(6.7, 3.3), new Point(6.1, 3.4));

        Iterable<Point> actualSkylines = bnl.computeSkylinesWithPreComparison(flagsPoints);

        assertThat(actualSkylines).containsOnlyElementsOf(expectedSkylines);
    }

    @Test public void computingSkylinesWithtPreComparison_withOnlySkylines_returnsTheSamePoints() {
        List<Tuple2<Flag, Point>> flagsPoints = Arrays.asList(
                new Tuple2<>(new Flag(true, false), new Point(5.0, 4.1)),
                new Tuple2<>(new Flag(true, false), new Point(5.9, 4.0)),
                new Tuple2<>(new Flag(true, false), new Point(6.7, 3.3)),
                new Tuple2<>(new Flag(true, false), new Point(6.1, 3.4)),
                new Tuple2<>(new Flag(false, true), new Point(2.5, 7.3)));
        List<Point> expectedSkylines = Arrays.asList(
                new Point(5.0, 4.1), new Point(5.9, 4.0), new Point(6.7, 3.3),
                new Point(2.5, 7.3), new Point(6.1, 3.4));

        Iterable<Point> actualSkylines = bnl.computeSkylinesWithPreComparison(flagsPoints);

        assertThat(actualSkylines).containsOnlyElementsOf(expectedSkylines);
    }
}